src/main/java/net/juniper/netconf/element/Hello.java

```diff
@@ -0,0 +1,165 @@
+package net.juniper.netconf.element;
+
+import lombok.Builder;
+import lombok.Data;
+import lombok.EqualsAndHashCode;
+import lombok.Singular;
+import lombok.ToString;
+import lombok.extern.slf4j.Slf4j;
+import net.juniper.netconf.NetconfConstants;
+import org.w3c.dom.Document;
+import org.w3c.dom.Element;
+import org.w3c.dom.Node;
+import org.w3c.dom.NodeList;
+import org.xml.sax.InputSource;
+import org.xml.sax.SAXException;
+
+import javax.xml.parsers.DocumentBuilderFactory;
+import javax.xml.parsers.ParserConfigurationException;
+import javax.xml.transform.OutputKeys;
+import javax.xml.transform.Transformer;
+import javax.xml.transform.TransformerException;
+import javax.xml.transform.TransformerFactory;
+import javax.xml.transform.dom.DOMSource;
+import javax.xml.transform.stream.StreamResult;
+import javax.xml.xpath.XPath;
+import javax.xml.xpath.XPathConstants;
+import javax.xml.xpath.XPathExpressionException;
+import javax.xml.xpath.XPathFactory;
+import java.io.IOException;
+import java.io.StringReader;
+import java.io.StringWriter;
+import java.util.List;
+
+/**
+ * Class to represent a NETCONF hello element - https://datatracker.ietf.org/doc/html/rfc6241#section-8.1
+ */
+@Data
+@Slf4j
+public class Hello {
+
+    @ToString.Exclude
+    @EqualsAndHashCode.Exclude
+    private final Document document;
+
+    @ToString.Exclude
+    private final String xml;
+
+    private final String sessionId;
+
+    @Singular("capability")
+    private final List<String> capabilities;
+
+    public boolean hasCapability(final String capability) {
+        return capabilities.contains(capability);
+    }
+
+    public String toXML() {
+        return xml;
+    }
+
+    /**
+     * Creates a Hello object based on the supplied XML.
+     *
+     * @param xml The XML of the NETCONF &lt;hello&gt;
+     * @return an new, immutable, Hello object.
+     * @throws ParserConfigurationException If the XML parser cannot be created
+     * @throws IOException                  If the XML cannot be read
+     * @throws SAXException                 If the XML cannot be parsed
+     * @throws XPathExpressionException     If there is a problem in the parsing expressions
+     */
+    public static Hello from(final String xml)
+            throws ParserConfigurationException, IOException, SAXException, XPathExpressionException {
+
+        final DocumentBuilderFactory documentBuilderFactory = DocumentBuilderFactory.newInstance();
+        documentBuilderFactory.setNamespaceAware(true);
+        final Document document = documentBuilderFactory.newDocumentBuilder()
+                .parse(new InputSource(new StringReader(xml)));
+        final XPath xPath = XPathFactory.newInstance().newXPath();
+        final String sessionId = xPath.evaluate("/*[namespace-uri()='urn:ietf:params:xml:ns:netconf:base:1.0' and local-name()='hello']/*[namespace-uri()='urn:ietf:params:xml:ns:netconf:base:1.0' and local-name()='session-id']", document);
```

Comment:
that is a super long string. Would it make sense to add this to the NetconfConstants.java?

Reply:
There’s certainly some refactoring that could be done here. However, I plan to do it with the <rpc-reply> implementation. My current thinking is that both Hello and RpcReply will inherit from AbstractNetconfElement  which will contain various utility methods that can be shared between them. So I’d like to put it off for now, if that’s OK.

Reply:
👍